squidpy/gr/_ppatterns.py

```diff
@@ -251,20 +234,54 @@ def _moran_helper(
     return pd.DataFrame(moran_list, columns=["I", "pval_sim", "VI_sim"], index=gen)
 
 
-def _compute_moran(y: np.ndarray, w: W, transformation: str, permutations: int) -> Tuple[float, float, float]:
-    mi = esda.moran.Moran(y, w, transformation=transformation, permutations=permutations)
-    return mi.I, mi.p_z_sim, mi.VI_sim
+@njit(
+    ft(ft[:], it[:], it[:], ft[:]),
+    parallel=False,
+    fastmath=True,
+)
+def _compute_moran(
+    counts: np.ndarray,
+    indptr: np.ndarray,
+    indices: np.ndarray,
+    data: np.ndarray,
+) -> Any:
+
+    z = counts - counts.mean()
+    zl = np.empty(len(z))
+    N = len(indptr) - 1
+    for i in range(N):
+        s = slice(indptr[i], indptr[i + 1])
+        i_indices = indices[s]
+        i_data = data[s]
+        zl[i] = np.sum(i_data * z[i_indices])
+    inum = (z * zl).sum()
+    z2ss = (z * z).sum()
+    return len(counts) / data.sum() * inum / z2ss
 
 
-def _set_weight_class(adata: AnnData, key: str) -> W:
-    X = adata.obsp[key]
-    if not isspmatrix_lil(X):
-        X = X.tolil()
+@njit(
+    ft[:](ft[:], it[:], it[:], ft[:], it),
```

Comment:
Indices of sparse matrices are not always `int32`, if you have enough entries they will become `int64`.

Reply:
Also, probably good to collect to `float64` if you're expecting a lot of samples.

Reply:
done, thank you!